appinventor/components/src/com/google/appinventor/components/runtime/CSVFile.java

```diff
@@ -0,0 +1,280 @@
+package com.google.appinventor.components.runtime;
+
+import com.google.appinventor.components.annotations.*;
+import com.google.appinventor.components.common.ComponentCategory;
+import com.google.appinventor.components.common.PropertyTypeConstants;
+import com.google.appinventor.components.common.YaVersion;
+import com.google.appinventor.components.runtime.util.*;
+
+import java.io.*;
+import java.util.*;
+import java.util.concurrent.*;
+
+@DesignerComponent(version = YaVersion.FILE_COMPONENT_VERSION,
+        description = "To be updated",
+        category = ComponentCategory.STORAGE,
+        nonVisible = true,
+        iconName = "images/file.png")
+@SimpleObject
+@UsesPermissions(permissionNames = "android.permission.WRITE_EXTERNAL_STORAGE, android.permission.READ_EXTERNAL_STORAGE")
+public class CSVFile extends FileBase {
+    private String sourceFile;
+
+    private YailList rows;
+    private YailList columns;
+    private YailList columnNames; // Elements of the first column
+
+    // private Thread readThread; // Async thread for parsing CSV
+    private ExecutorService threadRunner;
+
+    /**
+     * Creates a new CSVFile component.
+     * @param container the Form that this component is contained in.
+     */
+    public CSVFile(ComponentContainer container) {
+        super(container);
+
+        rows = new YailList();
+        columns = new YailList();
+
+        threadRunner = Executors.newSingleThreadExecutor();
+    }
+
+    /**
+     * Rows property getter method
+     *
+     * @return a YailList representing the parsed rows of the CSV file.
+     */
+    @SimpleProperty(category = PropertyCategory.BEHAVIOR, description = "Returns a list of rows corresponding" +
+            " to the CSV file's content.")
+    public YailList Rows() {
+        return getCSVPropertyHelper(new Callable<YailList>() {
+            @Override
+            public YailList call() throws Exception {
+                return rows;
+            }
+        });
+    }
+
+
+    /**
+     * Columns property getter method
+     *
+     * @return a YailList representing the parsed columns of the CSV file.
+     */
+    @SimpleProperty(category = PropertyCategory.BEHAVIOR, description = "Returns a list of columns corresponding" +
+            " to the CSV file's content.")
+    public YailList Columns() {
+        return getCSVPropertyHelper(new Callable<YailList>() {
+            @Override
+            public YailList call() throws Exception {
+                return columns;
```

Comment:
Try putting Callable in helper method (although it might not work due to final constraint)